### PR TITLE
refactor: Migrate EmojiInputViewModel to Typescript

### PR DESCRIPTION
**Pull request by thisisamir98**

### CODE REVIEW

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 1]
```diff
@@ -27,8 +29,26 @@ import {sortByPriority} from 'Util/StringUtil';
 import emojiBindings from './emoji.json';
 import {PROPERTIES_TYPE} from '../../properties/PropertiesType';
 import {StorageKey} from '../../storage/StorageKey';
+import {PropertiesRepository} from 'src/script/properties/PropertiesRepository';
+
+type EmojiData = {
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]:
```suggestion
export interface EmojiData {
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 2]
```diff
@@ -27,8 +29,26 @@ import {sortByPriority} from 'Util/StringUtil';
...
+  aliases: string[];
+  code: string;
+  name: string;
+};
+
+type EmojiBinding = {[key in string]: EmojiData};
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]:
```suggestion
type EmojiBinding = Record<string, EmojiData>;
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 3]
```diff
@@ -27,8 +29,26 @@ import {sortByPriority} from 'Util/StringUtil';
...
 
 export class EmojiInputViewModel {
+  private isVisible: boolean;
+  private readonly emojiList: {icon: string; name: string}[];
+  private emojiDict: {[key in string]: string} = {};
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]:
```suggestion
  private emojiDict: Record<string, string> = {};
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 4]
```diff
@@ -27,8 +29,26 @@ import {sortByPriority} from 'Util/StringUtil';
...
+  private readonly emojiDiv: JQuery<HTMLElement>;
+  private emojiStartPosition: number;
+  private emojiUsageCount: {[key in string]: number} = {};
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]:
```suggestion
  private emojiUsageCount: Record<string, number> = {};
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 5]
```diff
@@ -42,7 +62,7 @@ export class EmojiInputViewModel {
   // DO NOT USE COLON WITH LOWERCASE LETTERS IN THE SHORTCUTS, or you will prevent searching emojis.
   // For example, while :D should be replaced with unicode symbol, :d should allow searching for :dancer:
   /* eslint-disable sort-keys-fix/sort-keys-fix, no-multi-spaces */
-  static get INLINE_REPLACEMENT() {
+  static get INLINE_REPLACEMENT(): {name: string; shortcut: string}[] {
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts:65]:
```suggestion
  static get INLINE_REPLACEMENT(): Array<{name: string; shortcut: string}> {
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 6]
```diff
@@ -223,65 +239,68 @@ export class EmojiInputViewModel {
...
         }
       }
     }
 
     return true;
-  }
+  };
 
-  _initSubscriptions() {
+  initSubscriptions = (): void => {
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]:
```suggestion
  private readonly initSubscriptions = (): void => {
```

File: src/script/view_model/content/EmojiInputViewModel.ts [hunk 7]
```diff
@@ -345,7 +364,7 @@ export class EmojiInputViewModel {
         }
 
         return queryWords.every(queryWord => {
-          return emojiNameWords.some(emojiNameWord => emojiNameWord.startsWith(queryWord));
+          return emojiNameWords.some((emojiNameWord: string) => emojiNameWord.startsWith(queryWord));
```
Comment (ffflorian) [inline src/script/view_model/content/EmojiInputViewModel.ts]: This should work:

```suggestion
          return emojiNameWords.some<string>(emojiNameWord => emojiNameWord.startsWith(queryWord));
```

File: src/script/view_model/content/EmojiInputViewModel.ts (hunk 5 above)
Comment (thisisamir98) [reply src/script/view_model/content/EmojiInputViewModel.ts:65]: this one gives a lint error to use [] instead of Array<>

Comment (thisisamir98) [reply src/script/view_model/content/EmojiInputViewModel.ts:65]: image

Comment (ffflorian) [reply src/script/view_model/content/EmojiInputViewModel.ts:65]: ok :)

**Pull request merged**